# 3.6 손글씨 숫자 인식

학습 과정은 생략하고, 추론 과정만 구현하기  
추론과정: 순전파(forward propagation)  

## 3.6.1 MNIST 데이터셋

In [9]:
import sys, os
sys.path.append(os.pardir) # 부모 디렉터리의 파일을 가져올 수 있도록 설정
from dataset.mnist import load_mnist

(X_train, t_train), (X_test, t_test) = load_mnist(flatten = True, normalize = False)

In [10]:
# 각 데이터의 형상 출력

print(X_train.shape)
print(t_train.shape)
print(X_test.shape)
print(t_test.shape)

(60000, 784)
(60000,)
(10000, 784)
(10000,)


In [11]:
from PIL import Image
import numpy as np

def img_show(img):
    pil_img = Image.fromarray(np.uint8(img))
    pil_img.show()

img = X_train[0]
label = t_train[0]
print(label)
print(img.shape)

print(img.shape)
img = img.reshape(28, 28)
print(img.shape)

img_show(img)

5
(784,)
(784,)
(28, 28)


## 3.6.2 신경망의 추론 정리

이미지의 크기가 28*28 = 784이므로 입력층은 784개  
출력층은 0~9까지를 구별하는 문제이기 때문에 10개  
은닉층은 50개, 100개의 뉴런을 배치하는 2개를 넣을 것이다. (여기서 50개, 100개는 임의로 정한 값이다. )

In [12]:
def get_data():
    (x_train, t_train), (x_test, t_test) = load_mnist(normalize=True, flatten=True, one_hot_label=False)
    return x_test, t_test


def init_network():
    with open("sample_weight.pkl", 'rb') as f:
        network = pickle.load(f)
    return network


def predict(network, x):
    W1, W2, W3 = network['W1'], network['W2'], network['W3']
    b1, b2, b3 = network['b1'], network['b2'], network['b3']

    a1 = np.dot(x, W1) + b1
    z1 = sigmoid(a1)
    a2 = np.dot(z1, W2) + b2
    z2 = sigmoid(a2)
    a3 = np.dot(z2, W3) + b3
    y = softmax(a3)

    return y

In [20]:
def sigmoid(x):
    return 1 / (1 + np.exp(-x)) 
    
def softmax(x):
    if x.ndim == 2:
        x = x.T
        x = x - np.max(x, axis=0)
        y = np.exp(x) / np.sum(np.exp(x), axis=0)
        return y.T    
    
x, t = get_data()
network = init_network()
accuracy_cnt = 0
for i in range(len(x)):
    y = predict(network, x[i])
    p= np.argmax(y) # 확률이 가장 높은 원소의 인덱스를 얻는다.
    if p == t[i]:
        accuracy_cnt += 1

print("Accuracy:" + str(float(accuracy_cnt) / len(x)))

Accuracy:0.098
